pageserve/test_routes.py

```diff
@@ -1,39 +1,154 @@
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-"""
+"""Unit tests for pageserve service HTTP routes."""
+
+# Authors: mukobi
+# Copyright 2019 The Knative Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
 
 import unittest
 from unittest.mock import patch, MagicMock
-from flask import Flask
+import ast
+import requests_mock
 from flask_testing import TestCase
+import flask
 import app
 
+VALID_SESSION = {
+    "user_id": "42",
+    "name": "Ford Prefect",
+    "gauth_token": "I am a pretend token"}
+ERROR_BAD_TOKEN_TEXT = "Error: bad gauth_token."
+ERROR_BAD_TOKEN_STATUS = 400
+VALID_USER_AUTH_STATUS = 201
+
 EXAMPLE_USER = "app_user"
 EXAMPLE_POSTS = ['example', 'posts', 'list']
 EXAMPLE_EVENTS = ['example', 'events', 'list']
 
 
+class TestAuthenticateAndGetUser(unittest.TestCase):
+    """Test authentication endpoint POST /v1/authenticate."""
+
+    def assert_sessions_are_equal(self, first, second):
+        """Asserts the two sessions are the same.
+
+        Only compares the "user_id" and "name" fields, but ignores the
+        "gauth_token" field which is not returned from the users service
+        but is stored in the session by app.authenticate_with_users_service.
+        """
+        self.assertEqual(first["user_id"], second["user_id"])
+        self.assertEqual(first["name"], second["name"])
+
+    def setUp(self):
+        """Set up test client."""
+        app.app.config["TESTING"] = True  # propagate exceptions to test client
+        app.app.secret_key = "Dummy key used for testing the flask session"
+
+    @requests_mock.Mocker()
+    def test_valid_authentication(self, requests_mocker):
+        """GAuth token successfully authenticates."""
+        with app.app.test_request_context(), app.app.test_client() as client:
+            requests_mocker.post(app.app.config["USERS_ENDPOINT"] + "authenticate",
+                                 json=VALID_SESSION,
+                                 status_code=VALID_USER_AUTH_STATUS)
+            result = client.post("/v1/authenticate", data={
+                "gauth_token": "I don't matter because requests is mocked"})
+            self.assertEqual(result.status_code, VALID_USER_AUTH_STATUS)
+            result_dict = ast.literal_eval(result.data.decode())
```

Comment:
Why do you need ast here when you're just decoding a dict? Especially since this is the only place you use it.

Reply:
JSON using double quotes for strings. `client.post()` returns a Python dict-like string, not JSON, that has single quoted keys, so `json.loads()` raises an exception on it. `ast.literal_eval()` here is used to decode the result into a Python dict correctly.